pageserver/src/tenant/storage_layer/layer.rs

```diff
@@ -0,0 +1,1150 @@
+    }
+}
+
+impl std::fmt::Debug for ResidentLayer {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", self.owner)
+    }
+}
+
+impl ResidentLayer {
+    /// Release the eviction guard, making this back into a plain [`Layer`].
+    pub(crate) fn drop_eviction_guard(self) -> Layer {
+        self.into()
+    }
+
+    /// Loads all keys stored in the layer. Returns key, lsn and value size.
+    pub(crate) async fn load_keys(
+        &self,
+        ctx: &RequestContext,
+    ) -> anyhow::Result<Vec<DeltaEntry<'_>>> {
+        use LayerKind::*;
+
+        let owner = &self.owner.0;
+
+        match self.downloaded.get(owner).await? {
+            Delta(d) => {
+                owner
+                    .access_stats
+                    .record_access(LayerAccessKind::KeyIter, ctx);
+
+                // this is valid because the DownloadedLayer::kind is a OnceCell, not a
+                // Mutex<OnceCell>, so we cannot go and deinitialize the value with OnceCell::take
+                // while it's being held.
+                d.load_keys().await.context("Layer index is corrupted")
+            }
+            Image(_) => anyhow::bail!("cannot load_keys on a image layer"),
+        }
+    }
+
+    pub(crate) fn local_path(&self) -> &Path {
+        &self.owner.0.path
+    }
+
+    pub(crate) fn access_stats(&self) -> &LayerAccessStats {
+        self.owner.access_stats()
+    }
+}
+
+impl AsLayerDesc for ResidentLayer {
+    fn layer_desc(&self) -> &PersistentLayerDesc {
+        self.owner.layer_desc()
+    }
+}
+
+impl AsRef<Layer> for ResidentLayer {
+    fn as_ref(&self) -> &Layer {
+        &self.owner
+    }
+}
+
+/// Allow slimming down if we don't want the `2*usize` with eviction candidates?
+impl From<ResidentLayer> for Layer {
+    fn from(value: ResidentLayer) -> Self {
+        value.owner
+    }
+}
+
+/// Holds the actual downloaded layer, and handles evicting the file on drop.
+pub(crate) struct DownloadedLayer {
+    owner: Weak<LayerInner>,
+    kind: tokio::sync::OnceCell<anyhow::Result<LayerKind>>,
+}
+
+impl std::fmt::Debug for DownloadedLayer {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("DownloadedLayer")
+            // FIXME: this is not useful, always "Weak"
+            .field("owner", &self.owner)
+            .field("kind", &self.kind)
+            .finish()
+    }
+}
+
+impl Drop for DownloadedLayer {
+    fn drop(&mut self) {
+        if let Some(owner) = self.owner.upgrade() {
+            owner.on_drop();
+        } else {
+            // no need to do anything, we are shutting down
+        }
+    }
+}
+
+impl DownloadedLayer {
+    /// The owner is required so that we don't have to upgrade the `Self::owner`, which will only
+    /// be used on drop. This way, initializing a DownloadedLayer with the owner gone is
+    /// impossible.
+    async fn get(&self, owner: &LayerInner) -> anyhow::Result<&LayerKind> {
```

Comment:
Last sentence doesn't make sense to me. 

And I think there should be a one-line sentence stating what this function does.
Followed by the current doc comment that explains why it's necessry to supply the `owner` in order to do what the function does.

That might help resolve my confusion about the last sentence as well.

Reply:
Also, can we somehow assert first line in this function that `self.owner == owner` in terms of memory location?

Reply:
> Last sentence doesn't make sense to me.

Quite small time ago I had:

```rust
async fn get(&self) -> anyhow::Result<&LayerKind> {
    let init = || async {
        let Some(owner) = self.owner.upgrade() else {
            anyhow::bail!("timeline or pageserver is shutting down");
        };
        
        ...
    };
    
    self.kind.get_or_init(init).await.as_ref().map_err(|e| anyhow!("layer loading failed: {e:#}")
}
```

So this does have the property that we can always (seemingly) fail with the initialization because owner is gone.

However, the only time we should end up calling this method is through `Layer::get_value_reconstruct_data` OR `ResidentLayer::load_keys`. Which both hold at least a Weak reference to the `DownloadedLayer` at rest, when they end up calling the actual methods, they will have a `&DownloadedLayer` and  `&LayerInner` at hand.

So, we can **get rid** of this error case by using what you've been using via "witnesses", except that here the witness is the actual would-be-upgraded value.

Because methods expecting `owner: &LayerInner` are not public, I don't see how you could end up in a situation where you could inject a different value here. If you were able, the loading should fail if the values copied over from `PersistentLayerDesc` were different -- otherwise they don't actually matter at all because the reference does not live longer than it's lifetime.

Reply:
Hmm if it was an `Arc<LayerInner>` it should be `::as_ptr` comparable with the `Weak<LayerInner>`. I'll try to add that. I already changed this earlier, will link the `assert!` commit.

Reply:
579e85e92